packages/apputils/src/vdom.ts

```diff
@@ -63,68 +141,117 @@ export abstract class VDomRenderer<
   /**
-   * Called to update the state of the widget.
-   *
-   * The default implementation of this method triggers
-   * VDOM based rendering by calling the this.render() method.
+   * Phosphor signal to connect to.
    */
-  protected onUpdateRequest(msg: Message): void {
-    let vnode = this.render();
-    if (Array.isArray(vnode)) {
-      ReactDOM.render(vnode, this.node);
-    } else {
-      ReactDOM.render<any>(vnode, this.node);
-    }
-  }
 
-  /* Called after the widget is attached to the DOM
+  signal: ISignal<SENDER, ARGS>;
+  /**
+   * Initial value to use for the sender, used before the signal emits a value.
+   * If not provided, initial sender will be undefined
+   */
+  sender?: SENDER;
+  /**
+   * Initial value to use for the args, used before the signal emits a value.
+   * If not provided, initial args will be undefined.
+   */
+  args?: ARGS;
+  /**
+   * Function mapping the last signal value or inital values to an element to render.
    *
-   * Make sure the widget is rendered, even if the model has not changed.
+   * Note: returns `React.ReactNode` as per
+   * https://github.com/sw-yx/react-typescript-cheatsheet#higher-order-componentsrender-props
    */
-  protected onAfterAttach(msg: Message): void {
-    MessageLoop.sendMessage(this, Widget.Msg.UpdateRequest);
-  }
 
+  children: (sender?: SENDER, args?: ARGS) => React.ReactNode;
   /**
-   * Render the content of this widget using the virtual DOM.
-   *
-   * This method will be called anytime the widget needs to be rendered,
-   * which includes layout triggered rendering and all model changes.
+   * Given the last signal value, should return whether to update the state or not.
    *
-   * Subclasses should define this method and use the current model state
-   * to create a virtual node or nodes to render.
+   * The default unconditionally returns `true`, so you only have to override if you want
+   * to skip some updates.
    */
-  protected abstract render():
-    | Array<React.ReactElement<any>>
-    | React.ReactElement<any>
-    | null;
 
-  private _model: T | null;
-  private _modelChanged = new Signal<this, void>(this);
+  shouldUpdate?: (sender: SENDER, args: ARGS) => boolean;
+}
+
+/**
+ * State for the UseSignal component
+ */
+export interface IUseSignalState<SENDER, ARGS> {
+  value: [SENDER, ARGS];
 }
 
 /**
- * Phosphor widget that renders React Element(s).
+ * UseSignal provides a way to hook up a Phosphor signal to a React element,
+ * so that the element is re-rendered every time the signal fires.
+ *
+ * It is implemented through the "render props" technique, using the `children`
+ * prop as a function to render, so that it can be used either as a prop or as a child
+ * of this element
+ * https://reactjs.org/docs/render-props.html
+ *
+ *
+ * Example as child:
+ *
+ * ```
+ * function LiveButton(isActiveSignal: ISignal<any, boolean>) {
+ *  return (
+ *    <UseSignal signal={isActiveSignal} initialArgs={True}>
```

Comment:
I think I like `initialArgs` and `initialSender` instead of `args` and `sender` (which the code is right now). I was confused for a while that I was overriding the sender and args in the child function, rather than just supplying an initial value. Another way to do it would be to do `initial = {sender, args}`, or even `initial = [sender, args]` so that initial is just the first arguments to pass to the children function.

Reply:
Woops, yeah I removed the `initial` prefix and forgot to update them.

I added back the `initial` prefix, so they are `initialSender` and `initialArgs`.

I tried doing `initial = [sender, args]` at first, but I think I like the separate ones, because components often use either the sender, or the args, or neither, so making them separate allows users to only set a default if they use it.

Reply:
I also played around with the different options, and I agree - I like the separate initialSender and initialArgs better right now.